### ISSUE TITLE
JupyterLab UI widgets not available to browser as links

### ISSUE THREAD

**Issue by dabro** [jupyterlab/jupyterlab#134]
I use Firefox with a plugin called [Vimperator](https://en.wikipedia.org/wiki/Vimperator), which allows quick keyboard-based web-page control.  There are similar plugins for Chrome (cVim, Vimium) and Safari.  The JupyterLab alpha page control elements (file list, command list, console tabs, etc) are not exposed as links, so are not available for rapid keyboard clickthrough (as far as I know), or for scripting/extension via Jupyter-agnostic browser extensions that look for refs, etc.

Here is a picture from Vimperator in Firefox/Jupyter:
![screen capture - lack of link access firefox](https://cloud.githubusercontent.com/assets/9028933/16246302/31bcae7e-37ca-11e6-9abd-53efdd186c5b.png)

And here from cVim in Chrome:
![screen capture - lack of link access chrome](https://cloud.githubusercontent.com/assets/9028933/16246301/31b62202-37ca-11e6-8ad5-e069a37730bb.png)

You can see that while the `workspace` frame is accessible via quicklink (can get to the filter text box), the individual files/commands are not.  Ideally, each of the listed files or commands visible on the screen would be visible to the browser as an href, for extensions like this to detect.  Same for the tabs, eg `About` and `JupyterLab`.  

These extensions allow a Vim-like keyboard-driven UI--I rarely reach for the mouse to navigate Firefox or Chrome, except for edge cases.  I am able to do all the address-bar, link navigation, search, page content selection/highlighting, and browser tab management from a "command-line", which is scriptable and extensible itself.  This simulates Vim's `ex`/command mode, or Emacs command line, from eg `M-x`.  Here's an example of pressing `f` in Vimperator on a notebook:
![screen capture - notebook in chrome with shortcuts highlighted](https://cloud.githubusercontent.com/assets/9028933/16248141/eefe5796-37d2-11e6-8f4a-c946a218460d.png)

I know it looks like a soup sandwich, but that is just because I have the link hints blown up.  When I escape to `normal mode` (`ESC` a la Vim) and type `f` (`F` if I want to open the link in a new browser tab), each available link is highlighted with a letter combination which allows quick selection and page navigation.  When I select an editable element, or I type `i` with a text box selected, I enter `insert mode` (a la Vim).

This enables an environment very comfortable to those who like the low-drag UI of legacy text editors.  Combined with the awesome work you've done, this actually enables a rough approximation of an IDE, which can be really usefull.  Below is a picture of a saved layout I have:
![screen capture - ghetto ide](https://cloud.githubusercontent.com/assets/9028933/16247344/eed023c0-37ce-11e6-8761-e8c093eb643a.png)
Obviously Jupyter Lab supersedes much of that, but the main point is that if you expose your page granularly to the host browser, people can extend your work in awesome ways, short of hacking the javascript.  I think a lot of your technical users are probably in the camp of being able to configure browser extensions to create customized mashups, but not navigate the specifics of your javascript and css design.  

The various extensions work in quite different ways, so short of html standards, there is no way (that I know of) to make your pages compatible.  In vimperator's specific case, it can access any of the javascript content on a page like the javascript console, so one could script a specific site via something like [getElementByID in Vimperator](http://stackoverflow.com/questions/15995201/how-do-i-getelementbyid-in-vimperator)

```
js window.content.window.document.getElementById("myDiv").style.color="blue";
```

So access like this can be plugged into javascript functions defined in a `vimrc`-like file (`.vimperatorrc`) to define custom actions, which can then be assigned to a specific key-combination for quick access:

```
" Run :vimperatortoggle to enable or disable Vimperator as needed.
javascript function vimperatortoggle() { 
   if (is_vimperator_active) { liberator.execute(':disable') } 
   else { liberator.execute(':enable') }; 
}
"creates vimperator command (accessed from `:`) `vimperatortoggle` which calls
"the javascript function
command! vimperatortoggle javascript vimperatortoggle()

" Run :disable to disable Vimperator and resume normal Firefox behavior.
javascript function disable() { 
    is_vimperator_active = false; 
    liberator.execute('set gui=addons,bookmarks,menu,navigation,tabs'); 
    modes.passAllKeys = true; 
}

command! disable javascript disable()

" Run :enable to enable Vimperator.
javascript function enable() { 
    is_vimperator_active = true; 
    liberator.execute('set gui=addons,nobookmarks,menu,navigation'); 
    modes.passAllKeys = false; 
}

command! enable javascript enable()

" Press 'C-q' to toggle Vimperator
noremap <C-q> :vimperatortoggle<CR>
```

This in turn can be packaged up in a "plugin" and distributed like vim plugins to other vimperator users.

**Comment by dabro** [comment]
To be clear, this Vim-like extension is different than `vim-binding`, which enables CloudMirror's vim-bindings _inside_ a cell and adds some additional keybindings to cell editing and notebook editing modes. Vimperator doesn't enable vim text editing; it emulates the overall control/navigation/command experience of Vim in a browser context.  In fact, if you edit a text field while using Vimperator, you have _no_ access to vim text editing commands unless you use Vimperator to automatically launch the text field in an external editor like Vim.  So the vim notebook extensions and vim browser extensions are highly complementary.

**Comment by blink1073** [comment]
Thank you for the very well thought out issue and use case.

**Comment by dabro** [comment]
I have no experience with accessibility UI design, and I have no idea how/if IO devices like braille readers handle modern javascript web-apps, so please take this last comment with skepticism, but I'd think that making your elements easily accessible via basic browser operations would be a boon.  I guess that could be as easily accomplished by granular accelerator key settings too, instead of/in addition to, hrefs, like you did in the notebook modal keyboard UI.

You guys have done a really awesome job.  I think you're one step away from replacing Emacs/Vim adequately (finally!) with your cell-oriented kernel architecture.  Finally delivering on Knuth's vision for literate programming in a vehicle accessible to the general population.

**Comment by fperez** [comment]
This is excellent feedback, and we should really consider this class of ideas to at least support great keyboard handling for advanced users.  Many thanks!

Furthermore, getting proper URLs for all elements is something we've talked about as part of supporting session management and other issues.  This is a great description of why we should consider these questions...

**Comment by fperez** [comment]
Some sort of URL-based access to each main activity is probably going to be mandatory at least for 1.0, and @dabro's discussion above is excellent context.  I would almost argue this is 0.90 material, but I know it's hard and may require a lot of thinking, so I don't want to be overly ambitious.

But if we have a way to access individual elements by URL, then it becomes easier to imagine a model for full session restoring (since a session becomes a collection of URLs, potentially with a bit of layout info if we want to preserve the view faithfully).

And at the very least, URL access is something that users are completely accustomed to from today's notebook, and I know people use it a lot (we've often told people to simply use their browsers' bookmark system to keep references to commonly used notebooks or locations in the filesystem if they want).

**Comment by fperez** [comment]
While technically this is an enhancement, I think it's a very high-priority one that could almost be considered a usability bug.

**Comment by ellisonbg** [comment]
Clearing out older issue. The URL part of this is being tracked in multiple other issues. Current work is going on in #1075. Closing to track there instead.
